core/unix/signal.c

```diff
@@ -6631,11 +6633,120 @@ execute_native_handler(dcontext_t *dcontext, int sig, sigframe_rt_t *our_frame)
     report_app_problem(dcontext, APPFAULT_FAULT, (byte *)sc->SC_XIP, (byte *)sc->SC_FP,
                        "\nSignal %d delivered to application handler.\n", sig);
 
-    copy_frame_to_stack(dcontext, info, sig, our_frame, (void *)xsp, false /*!pending*/);
+    /*
+     * For a signal that arrived when the thread is native and the process is in
+     * the middle of detaching, we should try to reuse the current DR signal
+     * frame to deliver it if possible. This is to avoid using up too much app
+     * stack, which has been seen to cause a stack overflow in the app (i#6814).
+     *
+     * Summary of relevant events when DR starts detaching:
+     * D: Detacher thread sends suspend signal to all other threads.
+     * O: each of the Other threads receive the suspend signal. Note that DR's
+     *    signal handling configuration blocks all signals except SIGSEGV and
+     *    the suspend signal. So at this point Other threads cannot receive
+     *    app signals. Other threads signal to the Detacher thread that they're
+     *    suspended and awaiting wakeup call to continue detaching (sig_detach).
+     * D: Detacher thread sets doing_detach to true. A side effect of this is
+     *    that get_thread_private_dcontext() will return NULL when any Other
+     *    thread tries to get its private dcontext.
+     * D: Detacher thread wakes up each Other thread, telling it to detach.
+     * O: Each Other thread wakes up, does sig_detach which reinstates the app
+     *    signal stack (if available) and the app's blocked signal set on the
+     *    signal frame (it's restored on sigreturn from DR's handler), lets the
+     *    Detacher thread know that it's done detaching and resumes native
+     *    execution. Note that at this point, DR's sigact config is still
+     *    installed (including main_signal_handler). When the Other thread
+     *    returns from the suspend signal, signals get unblocked automatically
+     *    and may now be delivered.
+     * <may be able to reuse current frame for signals delivered here>
+     * D: When all threads are done detaching, the Detacher thread cleans up
+     *    each thread's drcontext for real, and also reinstates the app's
+     *    sigact in signal_thread_exit (which removes main_signal_handler).
+     *
+     * Therefore, if we arrive here and find doing_detach to be true and
+     * dcontext to be NULL, we can be certain that the current thread has
+     * completed sig_detach. This means the interrupted context stack (sc->XSP)
+     * must be the app's own stack.
+     *
+     * Also, if the thread has a signal stack at all for any signal, we will
+     * be on it (because DR sets SA_ONSTACK for all signals, and that config is
+     * still in effect), regardless of whether the signal stack was registered
+     * for sig.
+     */
+    bool reuse_cur_frame = false;
+    if (dcontext == NULL && doing_detach) {
+        ASSERT(cur_xsp != NULL);
```

Comment:
Above it says it's ok to pass NULL: maybe reword to more strongly say it's not ok for this case.